lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp

```diff
@@ -114,6 +115,74 @@ struct EnableGroupingPattern : public OpRewritePattern<ClockTreeOp> {
     return success(changed);
   }
 };
+
+struct GroupAssignmentsInIfPattern : public OpRewritePattern<scf::IfOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    // Pull values only used in certain reset/enable cases into the appropriate
+    // IfOps
+    // Skip anything not in a ClockTreeOp
+    if (auto clockTreeOp = ifOp->getParentOfType<ClockTreeOp>()) {
+      Region *groupingRegions[2] = {&ifOp.getThenRegion(),
+                                    &ifOp.getElseRegion()};
+      bool changed = false;
+      for (auto *region : groupingRegions) {
+        if (region->empty())
+          continue;
+
+        // Since we only work with IfOp then/else regions, we have at most 1
+        // block
+
+        Block &block = region->front();
+        SmallVector<Operation *> worklist;
+        // Don't walk as we don't want nested ops in order to restrict to IfOps
+        for (auto &op : block.getOperations()) {
+          worklist.push_back(&op);
+        }
+        while (!worklist.empty()) {
+          SmallVector<Operation *> theseOperands;
+          Operation *op = worklist.back();
+          for (auto operand : op->getOperands()) {
+            if (Operation *definition = operand.getDefiningOp()) {
+              // Skip if the operand is already defined in this block or is
+              // defined out of the clock tree
+              if (definition->getBlock() == op->getBlock() ||
+                  !clockTreeOp->isAncestor(definition))
+                continue;
+              if (!operand.hasOneUse()) {
+                bool safeToMove = true;
+                for (auto *user : operand.getUsers()) {
+                  if (!op->getParentRegion()->isAncestor(
+                          user->getParentRegion()) ||
+                      (user->getBlock() == op->getBlock() &&
+                       user->isBeforeInBlock(op))) {
+                    safeToMove = false;
+                    break;
+                  }
+                }
+                if (!safeToMove)
+                  break;
```

Comment:
Can't we just `continue` here?

Reply:
Ah, so we can, I think I restructured this and missed that! Thanks!